packages/core/crates/core-crypto/src/types.rs

```diff
@@ -0,0 +1,1016 @@
+        Self::deserialize(&hex::decode(s).map_err(|_| ParseError)?)
+    }
+}
+
+/// Represents a 256-bit hash value
+/// This implementation instantiates the hash via Keccak256 digest.
+#[derive(Clone, Eq, PartialEq, Debug)]
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+pub struct Hash {
+    hash: [u8; Self::SIZE],
+}
+
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+impl Hash {
+    #[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen(constructor))]
+    pub fn new(hash: &[u8]) -> Self {
+        assert_eq!(hash.len(), Self::SIZE, "invalid length");
+        let mut ret = Hash {
+            hash: [0u8; Self::SIZE]
+        };
+        ret.hash.copy_from_slice(hash);
+        ret
+    }
+}
+
+impl BinarySerializable for Hash {
+    const SIZE: usize = 32; // Defined by Keccak256.
+
+    fn deserialize(data: &[u8]) -> utils_types::errors::Result<Self> {
+        if data.len() == Self::SIZE {
+            let mut ret = Self {
+                hash: [0u8; Self::SIZE]
+            };
+            ret.hash.copy_from_slice(data);
+            Ok(ret)
+        } else {
+            Err(ParseError)
+        }
+    }
+
+    fn serialize(&self) -> Box<[u8]> {
+        self.hash.into()
+    }
+}
+
+impl Hash {
+    /// Takes all the byte slices and computes hash of their concatenated value.
+    /// Uses the Keccak256 digest.
+    pub fn create(inputs: &[&[u8]]) -> Self {
+        let mut hash = Keccak256::default();
+        inputs.into_iter().for_each(|v| hash.update(*v));
+        let mut ret = Hash {
+            hash: [0u8; Self::SIZE]
+        };
+        hash.finalize_into(&mut ret.hash).unwrap();
+        ret
+    }
+}
+
+/// Represents a secp256k1 public key.
+/// This is a "Schrödinger public key", both compressed and uncompressed to save some cycles.
+#[derive(Clone, Eq, PartialEq, Debug)]
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+pub struct PublicKey {
+    key: elliptic_curve::PublicKey<Secp256k1>,
+    compressed: Box<[u8]>
+}
+
+#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
+impl PublicKey {
+    pub fn to_address(&self) -> Address {
+        let uncompressed = self.serialize(false);
+        let serialized = Hash::create(&[&uncompressed[1..]]).serialize();
+        Address::new(&serialized[12..])
+    }
+
+    pub fn serialize(&self, compressed: bool) -> Box<[u8]> {
+        if compressed {
+            self.compressed.clone()
+        } else {
+            self.key.as_affine().to_encoded_point(false).to_bytes()
+        }
+    }
+
+    pub fn to_hex(&self, compressed: bool) -> String {
+        hex::encode(self.serialize(compressed))
+    }
+}
+
+impl PeerIdLike for PublicKey {
+    fn from_peerid(peer_id: &PeerId) -> utils_types::errors::Result<Self> {
+        // Workaround for the missing public key API on PeerIds
+        let peer_id_str = peer_id.to_base58();
+        if peer_id_str.starts_with("16U") {
```

Comment:
```suggestion
        // FIXME: use byte representation rather than string comparison
        if peer_id_str.starts_with("16U") {
```

Reply:
Yeah, it already says in the code comment above that this is currently a workaround for the missing pubkey type public API